File: surfaces/ruby/lib/b1/canon.rb
# frozen_string_literal: true

# B1-CANON-1 for Ruby. Normative definition: SPEC/20-b1-canon-1.md.
#
# JSON.parse is not used: it keeps the last of a set of duplicate member names, which is one of the
# silent repairs this profile exists to reject. The parser here is strict in the same way every
# other implementation is, which is what makes the corpus a real cross-check rather than a
# formality.

require 'digest'

module B1
  module Canon
    MAX_DEPTH = 64
    MAX_SAFE = 9_007_199_254_740_991
    KEY_RE = /\A[A-Za-z0-9_$.\-]{1,64}\z/
    ZERO_LINK = "b1c1:#{'0' * 64}"

    class Error < StandardError
      attr_reader :token

      def initialize(token, detail = '')
        @token = token
        super(detail.empty? ? token : "#{token}: #{detail}")
      end
    end

    # Distinguishes an object from an array, which a bare Ruby Hash/Array pair already does — but
    # the wrapper keeps the parsed shape explicit and stops a Hash with symbol keys leaking in.
    Obj = Struct.new(:members) do
      def get(key) = members[key]
    end

    class Parser
      def initialize(text)
        @s = text
        @i = 0
        @n = text.length
      end

      def parse
        ws
        v = value(0)
        ws
        raise Error.new('B1_ERR_PARSE', 'trailing input') unless @i == @n

        v
      end

      private

      def ws
        @i += 1 while @i < @n && [' ', "\t", "\n", "\r"].include?(@s[@i])
      end

      def literal(word)
        raise Error.new('B1_ERR_PARSE', "expected #{word}") unless @s[@i, word.length] == word

        @i += word.length
      end

      def value(depth)
        raise Error.new('B1_ERR_DEPTH', "depth > #{MAX_DEPTH}") if depth > MAX_DEPTH
        raise Error.new('B1_ERR_PARSE', 'unexpected end of input') if @i >= @n

        case @s[@i]
        when '{' then object(depth)
        when '[' then array(depth)
        when '"' then string
        when 't' then literal('true') || true
        when 'f' then literal('false') || false
        when 'n' then literal('null') || nil
        when '-', '0'..'9' then number
        else raise Error.new('B1_ERR_PARSE', "unexpected character #{@s[@i].inspect}")
        end
      end

      def object(depth)
        @i += 1
        members = {}
        ws
        if @s[@i] == '}'
          @i += 1
          return Obj.new(members)
        end

        loop do
          ws
          raise Error.new('B1_ERR_PARSE', 'expected key') unless @s[@i] == '"'

          key = string
          raise Error.new('B1_ERR_KEY_SYNTAX', key) unless key.match?(KEY_RE)
          raise Error.new('B1_ERR_DUPLICATE_KEY', key) if members.key?(key)

          ws
          raise Error.new('B1_ERR_PARSE', "expected ':'") unless @s[@i] == ':'

          @i += 1
          ws
          members[key] = value(depth + 1)
          ws
          case @s[@i]
          when ',' then @i += 1
          when '}' then (@i += 1) && (return Obj.new(members))
          else raise Error.new('B1_ERR_PARSE', "expected ',' or '}'")
          end
        end
      end

      def array(depth)
        @i += 1
        items = []
        ws
        if @s[@i] == ']'
          @i += 1
          return items
        end

        loop do
          ws
          items << value(depth + 1)
          ws
          case @s[@i]
          when ',' then @i += 1
          when ']' then (@i += 1) && (return items)
          else raise Error.new('B1_ERR_PARSE', "expected ',' or ']'")
          end
        end
      end

      def number
        start = @i
        @i += 1 if @s[@i] == '-'
        digits_start = @i
        @i += 1 while @i < @n && @s[@i] >= '0' && @s[@i] <= '9'
        raise Error.new('B1_ERR_PARSE', 'expected digits') if @i == digits_start
        if @i - digits_start > 1 && @s[digits_start] == '0'
          raise Error.new('B1_ERR_PARSE', 'leading zero')
        end
        if ['.', 'e', 'E'].include?(@s[@i])
          raise Error.new('B1_ERR_NONINTEGER_NUMBER', 'non-integer')
        end

        text = @s[start...@i]
        raise Error.new('B1_ERR_NONINTEGER_NUMBER', 'negative zero') if text == '-0'

        v = Integer(text, 10)
        if v > MAX_SAFE || v < -MAX_SAFE
          raise Error.new('B1_ERR_NONINTEGER_NUMBER', "out of range: #{text}")
        end

        v
      end

      def hex4
        hex = @s[@i, 4]
        raise Error.new('B1_ERR_PARSE', 'bad \\u escape') unless hex&.match?(/\A[0-9a-fA-F]{4}\z/)

        @i += 4
        hex.to_i(16)
      end

      # Joins a surrogate pair into one scalar; either half alone is rejected (SPEC/20 R3).
      def unicode_escape
        cp = hex4
        if cp >= 0xD800 && cp <= 0xDBFF
          unless @s[@i] == '\\' && @s[@i + 1] == 'u'
            raise Error.new('B1_ERR_INVALID_UTF8', 'unpaired high surrogate')
          end

          @i += 2
          low = hex4
          unless low >= 0xDC00 && low <= 0xDFFF
            raise Error.new('B1_ERR_INVALID_UTF8', 'high surrogate without a low one')
          end

          return (0x10000 + ((cp - 0xD800) << 10) + (low - 0xDC00)).chr(Encoding::UTF_8)
        end
        if cp >= 0xDC00 && cp <= 0xDFFF
          raise Error.new('B1_ERR_INVALID_UTF8', 'unpaired low surrogate')
        end

        cp.chr(Encoding::UTF_8)
      end

      def string
        @i += 1
        out = +''
        loop do
          raise Error.new('B1_ERR_PARSE', 'unterminated string') if @i >= @n

          c = @s[@i]
          if c == '"'
            @i += 1
            return out
          end

          if c == '\\'
            @i += 1
            e = @s[@i]
            raise Error.new('B1_ERR_PARSE', 'unterminated escape') if e.nil?

            @i += 1
            out << case e
                   when '"' then '"'
                   when '\\' then '\\'
                   when '/' then '/'
                   when 'b' then "\b"
                   when 'f' then "\f"
                   when 'n' then "\n"
                   when 'r' then "\r"
                   when 't' then "\t"
                   when 'u' then unicode_escape
                   else raise Error.new('B1_ERR_PARSE', "bad escape \\#{e}")
                   end
            next
          end

          raise Error.new('B1_ERR_PARSE', 'raw control character') if c.ord < 0x20

          out << c
          @i += 1
        end
      end
    end

    SHORT_ESCAPES = {
      '"' => '\\"', '\\' => '\\\\', "\b" => '\\b', "\t" => '\\t',
      "\n" => '\\n', "\f" => '\\f', "\r" => '\\r'
    }.freeze

    module_function

    def escape(str)
      out = +'"'
      str.each_char do |ch|
        short = SHORT_ESCAPES[ch]
        if short
          out << short
        elsif ch.ord < 0x20
          out << format('\\u%04x', ch.ord) # lowercase, per R3
        else
          out << ch
        end
      end
      out << '"'
    end

    def canonicalize(value, depth = 0)
      raise Error.new('B1_ERR_DEPTH', "depth > #{MAX_DEPTH}") if depth > MAX_DEPTH

      case value
      when nil then 'null'
      when true then 'true'
      when false then 'false'
      when Integer
        if value > MAX_SAFE || value < -MAX_SAFE
          raise Error.new('B1_ERR_NONINTEGER_NUMBER', 'out of range')
        end

        value.to_s
      when Float
        raise Error.new('B1_ERR_NONINTEGER_NUMBER', 'floating point is not representable')
      when String then escape(value)
      when Array
        "[#{value.map { |v| canonicalize(v, depth + 1) }.join(',')}]"
      when Obj then canonicalize_members(value.members, depth)
      when Hash then canonicalize_members(value, depth)
      else raise Error.new('B1_ERR_PARSE', "unsupported type #{value.class}")
      end
    end

    def canonicalize_members(members, depth)
      # R2 restricts names to ASCII, so a byte-order sort is the canonical order and means the same
      # thing in all fourteen languages.
      keys = members.keys.map(&:to_s).sort
      parts = keys.map do |k|
        raise Error.new('B1_ERR_KEY_SYNTAX', k) unless k.match?(KEY_RE)

        value = members.key?(k) ? members[k] : members[k.to_sym]
        "#{escape(k)}:#{canonicalize(value, depth + 1)}"
      end
      "{#{parts.join(',')}}"
    end

    def parse(text)
      Parser.new(text).parse
    end

    def digest_value(value)
      Digest::SHA256.hexdigest(canonicalize(value))
    end

    def digest_text(text)
      digest_value(parse(text))
    end

    # The prefix labels the algorithm and is not part of the hashed input.
    def b1c1(digest_hex) = "b1c1:#{digest_hex}"
  end
end

File: surfaces/php/canon.php
<?php

declare(strict_types=1);

/**
 * B1-CANON-1 for PHP. Normative definition: SPEC/20-b1-canon-1.md.
 *
 * json_decode is not used: it keeps the last of a set of duplicate member names, and PHP arrays
 * cannot distinguish an empty object from an empty array. Both would silently change a document
 * this profile is required to reject or to digest exactly.
 */

final class B1Exception extends RuntimeException
{
    public function __construct(public readonly string $token, string $detail = '')
    {
        parent::__construct($detail === '' ? $token : "$token: $detail");
    }
}

/** Distinguishes an object from an array, which a bare PHP array cannot. */
final class B1Obj
{
    /** @var array<string, mixed> */
    public array $members = [];

    public function get(string $key): mixed
    {
        return $this->members[$key] ?? null;
    }

    public function set(string $key, mixed $value): self
    {
        $this->members[$key] = $value;
        return $this;
    }
}

final class B1Arr
{
    /** @param list<mixed> $items */
    public function __construct(public array $items = [])
    {
    }
}

const B1_MAX_DEPTH = 64;
const B1_MAX_SAFE = 9007199254740991;

function b1_key_syntax_ok(string $k): bool
{
    return (bool) preg_match('/^[A-Za-z0-9_$.\-]{1,64}$/', $k);
}

final class B1Parser
{
    private int $i = 0;
    /** @var list<string> */
    private array $chars;

    public function __construct(private readonly string $s)
    {
        // Operate on code points: byte indexing would split multi-byte characters inside strings.
        $chars = preg_split('//u', $s, -1, PREG_SPLIT_NO_EMPTY);
        if ($chars === false) {
            throw new B1Exception('B1_ERR_INVALID_UTF8', 'input is not valid UTF-8');
        }
        $this->chars = $chars;
    }

    public function parse(): mixed
    {
        $this->ws();
        $v = $this->value(0);
        $this->ws();
        if ($this->i !== count($this->chars)) {
            throw new B1Exception('B1_ERR_PARSE', 'trailing input');
        }
        return $v;
    }

    private function ws(): void
    {
        while ($this->i < count($this->chars)
            && in_array($this->chars[$this->i], [' ', "\t", "\n", "\r"], true)) {
            $this->i++;
        }
    }

    private function literal(string $word): void
    {
        $chars = str_split($word);
        foreach ($chars as $k => $c) {
            if (($this->chars[$this->i + $k] ?? null) !== $c) {
                throw new B1Exception('B1_ERR_PARSE', "expected $word");
            }
        }
        $this->i += count($chars);
    }

    private function value(int $depth): mixed
    {
        if ($depth > B1_MAX_DEPTH) {
            throw new B1Exception('B1_ERR_DEPTH', 'depth > ' . B1_MAX_DEPTH);
        }
        if ($this->i >= count($this->chars)) {
            throw new B1Exception('B1_ERR_PARSE', 'unexpected end of input');
        }
        $c = $this->chars[$this->i];
        return match (true) {
            $c === '{' => $this->object($depth),
            $c === '[' => $this->array($depth),
            $c === '"' => $this->string(),
            $c === 't' => $this->boolLiteral('true', true),
            $c === 'f' => $this->boolLiteral('false', false),
            $c === 'n' => $this->nullLiteral(),
            $c === '-' || ($c >= '0' && $c <= '9') => $this->number(),
            default => throw new B1Exception('B1_ERR_PARSE', "unexpected character $c"),
        };
    }

    private function boolLiteral(string $word, bool $value): bool
    {
        $this->literal($word);
        return $value;
    }

    private function nullLiteral(): mixed
    {
        $this->literal('null');
        return null;
    }

    private function object(int $depth): B1Obj
    {
        $this->i++;
        $obj = new B1Obj();
        $this->ws();
        if (($this->chars[$this->i] ?? null) === '}') {
            $this->i++;
            return $obj;
        }
        while (true) {
            $this->ws();
            if (($this->chars[$this->i] ?? null) !== '"') {
                throw new B1Exception('B1_ERR_PARSE', 'expected key');
            }
            $key = $this->string();
            if (!b1_key_syntax_ok($key)) {
                throw new B1Exception('B1_ERR_KEY_SYNTAX', $key);
            }
            if (array_key_exists($key, $obj->members)) {
                throw new B1Exception('B1_ERR_DUPLICATE_KEY', $key);
            }
            $this->ws();
            if (($this->chars[$this->i] ?? null) !== ':') {
                throw new B1Exception('B1_ERR_PARSE', "expected ':'");
            }
            $this->i++;
            $this->ws();
            $obj->members[$key] = $this->value($depth + 1);
            $this->ws();
            $c = $this->chars[$this->i] ?? null;
            if ($c === ',') {
                $this->i++;
                continue;
            }
            if ($c === '}') {
                $this->i++;
                return $obj;
            }
            throw new B1Exception('B1_ERR_PARSE', "expected ',' or '}'");
        }
    }

    private function array(int $depth): B1Arr
    {
        $this->i++;
        $items = [];
        $this->ws();
        if (($this->chars[$this->i] ?? null) === ']') {
            $this->i++;
            return new B1Arr($items);
        }
        while (true) {
            $this->ws();
            $items[] = $this->value($depth + 1);
            $this->ws();
            $c = $this->chars[$this->i] ?? null;
            if ($c === ',') {
                $this->i++;
                continue;
            }
            if ($c === ']') {
                $this->i++;
                return new B1Arr($items);
            }
            throw new B1Exception('B1_ERR_PARSE', "expected ',' or ']'");
        }
    }

    private function number(): int
    {
        $start = $this->i;
        if ($this->chars[$this->i] === '-') {
            $this->i++;
        }
        $digitsStart = $this->i;
        while ($this->i < count($this->chars)
            && $this->chars[$this->i] >= '0' && $this->chars[$this->i] <= '9') {
            $this->i++;
        }
        if ($this->i === $digitsStart) {
            throw new B1Exception('B1_ERR_PARSE', 'expected digits');
        }
        if ($this->i - $digitsStart > 1 && $this->chars[$digitsStart] === '0') {
            throw new B1Exception('B1_ERR_PARSE', 'leading zero');
        }
        $next = $this->chars[$this->i] ?? null;
        if ($next === '.' || $next === 'e' || $next === 'E') {
            throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', 'non-integer');
        }
        $text = implode('', array_slice($this->chars, $start, $this->i - $start));
        if ($text === '-0') {
            throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', 'negative zero');
        }
        // Compare as a string before casting: an out-of-range literal would otherwise saturate or
        // wrap silently, turning a rejection into a wrong digest.
        $magnitude = ltrim($text, '-');
        if (strlen($magnitude) > 16
            || (strlen($magnitude) === 16 && bccomp_fallback($magnitude, (string) B1_MAX_SAFE) > 0)) {
            throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', "out of range: $text");
        }
        $v = (int) $text;
        if ($v > B1_MAX_SAFE || $v < -B1_MAX_SAFE) {
            throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', "out of range: $text");
        }
        return $v;
    }

    private function hex4(): int
    {
        $hex = implode('', array_slice($this->chars, $this->i, 4));
        if (strlen($hex) !== 4 || !ctype_xdigit($hex)) {
            throw new B1Exception('B1_ERR_PARSE', 'bad \\u escape');
        }
        $this->i += 4;
        return (int) hexdec($hex);
    }

    /** Joins a surrogate pair into one scalar; either half alone is rejected (SPEC/20 R3). */
    private function unicodeEscape(): string
    {
        $cp = $this->hex4();
        if ($cp >= 0xD800 && $cp <= 0xDBFF) {
            if (($this->chars[$this->i] ?? null) !== '\\' || ($this->chars[$this->i + 1] ?? null) !== 'u') {
                throw new B1Exception('B1_ERR_INVALID_UTF8', 'unpaired high surrogate');
            }
            $this->i += 2;
            $low = $this->hex4();
            if ($low < 0xDC00 || $low > 0xDFFF) {
                throw new B1Exception('B1_ERR_INVALID_UTF8', 'high surrogate without a low one');
            }
            return mb_chr(0x10000 + (($cp - 0xD800) << 10) + ($low - 0xDC00), 'UTF-8');
        }
        if ($cp >= 0xDC00 && $cp <= 0xDFFF) {
            throw new B1Exception('B1_ERR_INVALID_UTF8', 'unpaired low surrogate');
        }
        return mb_chr($cp, 'UTF-8');
    }

    private function string(): string
    {
        $this->i++;
        $out = '';
        while (true) {
            if ($this->i >= count($this->chars)) {
                throw new B1Exception('B1_ERR_PARSE', 'unterminated string');
            }
            $c = $this->chars[$this->i];
            if ($c === '"') {
                $this->i++;
                return $out;
            }
            if ($c === '\\') {
                $this->i++;
                $e = $this->chars[$this->i] ?? throw new B1Exception('B1_ERR_PARSE', 'unterminated escape');
                $this->i++;
                $out .= match ($e) {
                    '"' => '"',
                    '\\' => '\\',
                    '/' => '/',
                    'b' => "\x08",
                    'f' => "\x0c",
                    'n' => "\n",
                    'r' => "\r",
                    't' => "\t",
                    'u' => $this->unicodeEscape(),
                    default => throw new B1Exception('B1_ERR_PARSE', "bad escape \\$e"),
                };
                continue;
            }
            if (strlen($c) === 1 && ord($c) < 0x20) {
                throw new B1Exception('B1_ERR_PARSE', 'raw control character');
            }
            $out .= $c;
            $this->i++;
        }
    }
}

/** Compares two non-negative decimal strings without needing the bcmath extension. */
function bccomp_fallback(string $a, string $b): int
{
    $a = ltrim($a, '0') ?: '0';
    $b = ltrim($b, '0') ?: '0';
    if (strlen($a) !== strlen($b)) {
        return strlen($a) <=> strlen($b);
    }
    return strcmp($a, $b);
}

function b1_escape(string $s): string
{
    $out = '"';
    foreach (preg_split('//u', $s, -1, PREG_SPLIT_NO_EMPTY) ?: [] as $ch) {
        $out .= match ($ch) {
            '"' => '\\"',
            '\\' => '\\\\',
            "\x08" => '\\b',
            "\t" => '\\t',
            "\n" => '\\n',
            "\x0c" => '\\f',
            "\r" => '\\r',
            default => (strlen($ch) === 1 && ord($ch) < 0x20)
                ? sprintf('\\u%04x', ord($ch)) // lowercase, per R3
                : $ch,
        };
    }
    return $out . '"';
}

function b1_canonicalize(mixed $v, int $depth = 0): string
{
    if ($depth > B1_MAX_DEPTH) {
        throw new B1Exception('B1_ERR_DEPTH', 'depth > ' . B1_MAX_DEPTH);
    }
    if ($v === null) {
        return 'null';
    }
    if (is_bool($v)) {
        return $v ? 'true' : 'false';
    }
    if (is_int($v)) {
        if ($v > B1_MAX_SAFE || $v < -B1_MAX_SAFE) {
            throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', 'out of range');
        }
        return (string) $v;
    }
    if (is_float($v)) {
        throw new B1Exception('B1_ERR_NONINTEGER_NUMBER', 'floating point is not representable');
    }
    if (is_string($v)) {
        return b1_escape($v);
    }
    if ($v instanceof B1Arr) {
        return '[' . implode(',', array_map(
            static fn ($item) => b1_canonicalize($item, $depth + 1),
            $v->items
        )) . ']';
    }
    if ($v instanceof B1Obj) {
        // R2 restricts names to ASCII, so a byte-order sort is the canonical order.
        $keys = array_keys($v->members);
        sort($keys, SORT_STRING);
        $parts = [];
        foreach ($keys as $k) {
            if (!b1_key_syntax_ok((string) $k)) {
                throw new B1Exception('B1_ERR_KEY_SYNTAX', (string) $k);
            }
            $parts[] = b1_escape((string) $k) . ':' . b1_canonicalize($v->members[$k], $depth + 1);
        }
        return '{' . implode(',', $parts) . '}';
    }
    throw new B1Exception('B1_ERR_PARSE', 'unsupported value type');
}

function b1_digest_value(mixed $v): string
{
    return hash('sha256', b1_canonicalize($v));
}

function b1_digest_text(string $text): string
{
    return b1_digest_value((new B1Parser($text))->parse());
}
